src/self-update/README.md

```diff
@@ -0,0 +1,99 @@
+
+3) Package Dependencies need to be solved using SW Management before self-update. This is since self-update will disable auto-dep solving to reduce additional risk during self-update.
+
+*Hint: Before executing self-update on devices in the field it shall be tested on a kind of golden device that equals the device(s) in the field in system-environment, configuration and thin-edge version.*
+
+
+# Execution Flow and involved Components
+
+![Sequence Diagram Update SW-list](images/self-update.drawio.svg)
+
+
+# Design Principles
+
+### Details about the SM Plugin
+
+* Thin-edge updates are managed by a specific software management module type named `tedge`.
+
+* The existing SM plugin "apt" will be used and extended to manage that module type "tedge". 
+  * Same executable will be used for both module types, so a softlink must be located in `/etc/tedge/sm-plugins/` named `tedge`, that points to the plugin’s executable „apt“.
+
+* As the plugin serves both module types "tedge" and "apt", the SM Agent will execute the plugin twice with the LIST command; once for each mode type. To distinguish packages listed by APT package manager per module type all packages prefixed with "tedge_" are assumed as module type "tedge".
+  * When the plugin is called with LIST command for module type "tedge", packages prefixed with "tedge_" must be reported.
+  * When the plugin is called with LIST command for module type "apt", all packages withoutprefix "tedge_" must be reported.
+
+* The agent must request updates on the `tedge` plugin using the `update-list` API of this plugin. So the plugin can process the entire update request without any further SM Agent action. This is required because the agent will be restarted by the plugin.
+
+* Before any package update for module type "tedge" starts, all needed installation resources for the entire update request must be downloaded (i.E. all \*.deb packages for APT). That is to get the plugin uncoupled from any potential network issue during update procedure.
+  * For all packages with some URL given in the update request the SM Agent manages download before plugin start, and provides local file paths to the plugin.<br/>
+    TODO: What happens if one package download fails in SM agent? Does plugin maybe need to check if all files are locally available?
+  * For all packages without any URL given in the update request the Package Manager (i.E. APT) will manage the download as part of the update. Therefore implementation of plugin command update-list calls APT separately for download and for installation.
+    * First all packages are downloaded by calling APT with command line option `--download-only`.
+    * When all packages were successfully downloaded packages are installed by calling APT again with option `--no-download`.
+    * If one or more downloads fail, the plugin will not install any package, but return with error.
+
```

Comment:
After the installation of individual tedge packages are complete, some more actions will have to performed to wire things together, right? Like restarting starting things in a certain order or some other "update logic" that can't be performed at an individual binary's maintainer script level but have to be done as a post-processing step after every required component is installed? Isn't that also a responsibility of the plugin?

We may not be able to define what that logic would be right now, as it may change from one version to another(i.e, the post-processing required for an update to 0.7 might be different from an update to 0.6. It might even change based on the already installed version). So, just saying that it probably is worth acknowledging the possibility of additional post-processing beyond what the maintainer scripts of individual debian packages can handle.

Reply:
> Isn't that also a responsibility of the plugin?

No, the plugin can not do that job. Since it is part of the current installed (old) version, and can never know what the new installed version needs. Imagine there are new services, services removed, or other things to be considered, only the new version can take care. So finally it must be part of the new package(s).

> We may not be able to define what that logic would be right now, as it may change from one version to another

Exactly thats what I mean above. That logic must come with the new version. It might be directly part of the installation logic, or managed by some tedge init command (that obviously depends on the technical detail of the update step to be done). However it must be part of the new package and shall be triggered by the installation logic.